libc/string_lenchr.c

```diff
@@ -0,0 +1,210 @@
+	/*Empty input*/
+	TEST_ASSERT_EQUAL_INT(0, strlen(""));
+}
+
+
+/*
+//////////////////////////////////////////////////////////////////////////////////////////////////////////
+*/
+
+
+TEST_GROUP(string_strnlen);
+
+
+TEST_SETUP(string_strnlen)
+{
+}
+
+
+TEST_TEAR_DOWN(string_strnlen)
+{
+}
+
+
+TEST(string_strnlen, strnlen_right_input)
+{
+
+	const char empty[] = "\0";
+	char msg[] = { 'c', 'h', 'r', 'i', 's', 't', 'o', 'p', 'h', 'e', 'r', '\0' };
+	char *foo = "foo\0";
+	char *bar = "bar\0";
+	char str[12];
+	strcpy(str, "TEXT ");
+	strcat(str, foo);
+	strcat(str, bar);
+	TEST_ASSERT_EQUAL_INT(8, strnlen("New Case", 10));
+	TEST_ASSERT_EQUAL_INT(0, strnlen(empty, 0));
+	TEST_ASSERT_EQUAL_INT(11, strnlen(msg, 20));
+	TEST_ASSERT_EQUAL_INT(11, strnlen(str, 20));
+}
+
+
+TEST(string_strnlen, strnlen_mix_input)
+{
+	/*Checking if some of input can change outcom*/
+	TEST_ASSERT_EQUAL_INT(16, strnlen("Something \n Here", 20));
+	TEST_ASSERT_EQUAL_INT(16, strnlen("Something \t Here", 20));
+	TEST_ASSERT_EQUAL_INT(16, strnlen("Something \e Here", 20));
+	TEST_ASSERT_EQUAL_INT(16, strnlen("Something \r Here", 20));
+	TEST_ASSERT_EQUAL_INT(16, strnlen("Something \b Here", 20));
+	TEST_ASSERT_EQUAL_INT(16, strnlen("Something \v Here", 20));
+	TEST_ASSERT_EQUAL_INT(16, strnlen("Something \f Here", 20));
+	TEST_ASSERT_EQUAL_INT(16, strnlen("Something \r Here", 20));
+	TEST_ASSERT_EQUAL_INT(25, strnlen("Something Here \\ and here", 50));
+	TEST_ASSERT_EQUAL_INT(41, strnlen("Something Here %d/%f/%x/%i/%c/%s and here", 50));
+}
+
+
+TEST(string_strnlen, strnlen_special_input)
+{
+	char bigstr[5000] = "a";
+	/*Really big string to count*/
+	for (int i = 0; i < PATH_MAX - 1; i++) {
+		strcat(bigstr, "A");
+	}
+
+	TEST_ASSERT_EQUAL_INT(PATH_MAX, strnlen(bigstr, PATH_MAX));
+
+	TEST_ASSERT_EQUAL_INT(0, strnlen(""
+									 "",
+								 20));
+
+	/*Checking cappability of reading not only visible sings*/
+	TEST_ASSERT_EQUAL_INT(51, strnlen("12344567890!@#$^&*()_+~{[;'./<>?:/{empty}|:{}[]<>?:", 100));
+	TEST_ASSERT_EQUAL_INT(40, strnlen("\1\2\3\4\5\6\7\11\12\13\14\15\16\17\18\19\20\21\22\23\24\25\26\27\28\29\30\31\32\33\34\35\36\37\38\39\40", 40));
+	/*Adding null signs before actual value*/
+	TEST_ASSERT_EQUAL_INT(0, strnlen("\0\0\0\0 1 \0\0 1 \0\0 1 \0", 15));
+	/*Out of ascii signs*/
+	TEST_ASSERT_EQUAL_INT(36, strnlen("❉❉⌨⌨⌨⌨⌨⌨⌨⌨❉❉", 40));
+	/*Empty input*/
+	TEST_ASSERT_EQUAL_INT(0, strnlen("", 1));
+
+
+	TEST_ASSERT_EQUAL_INT(2, strnlen("New Case", 2));
+	/*Passing big number as numbers of elements*/
+	TEST_ASSERT_EQUAL_INT(8, strnlen("New Case", INT_MAX));
+	/*Hex representation of 144470*/
+	TEST_ASSERT_EQUAL_INT(8, strnlen("New Case", 0x23456));
+	/*Hex representation of 7*/
+	TEST_ASSERT_EQUAL_INT(7, strnlen("New Case", 0x7));
+	/*\3 and 99*/
+	TEST_ASSERT_EQUAL_INT(3, strnlen("\399\0", 6));
+}
+
+
+/*
+//////////////////////////////////////////////////////////////////////////////////////////////////
+*/
+
+TEST_GROUP_RUNNER(string_strlen)
```

Comment:
I see that the pattern is a little bit different than in previous test, but personally I like it more, because we see all groups' declarations at the beginning of the file and all cases listed at the end. If we decided to keep it, it would be good to change other tests also. Let's wait for the others' opinion.